crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -1080,6 +1082,15 @@ impl CheckpointSignatureAggregator {
         let author = signature.authority;
         // consensus ensures that authority == narwhal_cert.author
         if their_digest != self.digest {
+            if let InsertResult::QuorumReached(data) =
+                self.failures.insert_generic(author, signature)
+            {
+                panic!("Checkpoint fork detected - f+1 validators submitted checkpoint digest at seq {} different from our digest {}. Validators with different digests: {:?}",
```

Comment:
Did you mean 2f+1

Reply:
No, this panics on f+1. The logic is following - when f+1 submitted digest different from our, then our digest will never get to 2f+1, and therefore we can panic. Does this make sense?

Reply:
Got it. Any reason to not just make `failures` a HashSet of authority names, and assert on its length >= f+1? This may be unlikely but it is possible to see f+1 digests that are different from self but are not all the same.
I am also curious how we would eventually recover from this state (assuming we will release a fix, but how to make them not panic right away)

Reply:
But we need to count stake(which stake agg does) and not number of nodes?
Even if f+1 digests are not the same, this is still a network fork (since no valid hash will get 2f+1 votes at this point)